File: benchmarks/float_compute.go
package main

import (
	"fmt"
	"math"
	"time"
)

const runs = 3

func sumInvSquares(n int64) float64 {
	var sum float64 = 0.0
	for i := int64(1); i <= n; i++ {
		fi := float64(i)
		sum += 1.0 / (fi * fi)
	}
	return sum
}

func leibnizPi(n int64) float64 {
	var sum float64 = 0.0
	var sign float64 = 1.0
	for i := int64(1); i <= n; i++ {
		denom := float64(i*2 - 1)
		sum += sign / denom
		sign = -sign
	}
	return sum * 4.0
}

func mandelbrotIter(cr, ci float64, maxIter int64) int64 {
	var zr, zi float64 = 0.0, 0.0
	for i := int64(0); i < maxIter; i++ {
		tr := zr*zr - zi*zi + cr
		zi = 2.0*zr*zi + ci
		zr = tr
		if zr*zr+zi*zi > 4.0 {
			return i
		}
	}
	return maxIter
}

func distance(x1, y1, x2, y2 float64) float64 {
	dx := x2 - x1
	dy := y2 - y1
	return math.Sqrt(dx*dx + dy*dy)
}

func trigSum(n int64) float64 {
	var sum float64 = 0.0
	for i := int64(0); i < n; i++ {
		x := float64(i) * 0.001
		sum += math.Sin(x) + math.Cos(x)
	}
	return sum
}

func main() {
	fmt.Println("=== Go Float Benchmarks ===")

	// Benchmark 1: Basel problem
	n := int64(10000000)
	sumInvSquares(n) // warmup
	var best int64 = 1 << 62
	var resultF float64
	for run := 0; run < runs; run++ {
		start := time.Now()
		r := sumInvSquares(n)
		elapsed := time.Since(start).Nanoseconds()
		if elapsed < best {
			best = elapsed
			resultF = r
		}
	}
	fmt.Printf("Basel sum(1/i^2, 1..%d):\n", n)
	fmt.Printf("  Time: %d ms\n", best/1000000)
	fmt.Printf("  Result: %.10f\n", resultF)

	// Benchmark 2: Leibniz pi
	n = 10000000
	leibnizPi(n) // warmup
	best = 1 << 62
	for run := 0; run < runs; run++ {
		start := time.Now()
		r := leibnizPi(n)
		elapsed := time.Since(start).Nanoseconds()
		if elapsed < best {
			best = elapsed
			resultF = r
		}
	}
	fmt.Printf("Leibniz pi(%d terms):\n", n)
	fmt.Printf("  Time: %d ms\n", best/1000000)
	fmt.Printf("  Result: %.10f\n", resultF)

	// Benchmark 3: Mandelbrot grid
	size := int64(200)
	// warmup
	{
		var t int64 = 0
		for y := int64(0); y < size; y++ {
			for x := int64(0); x < size; x++ {
				cr := (float64(x) - 100.0) / 50.0
				ci := (float64(y) - 100.0) / 50.0
				t += mandelbrotIter(cr, ci, 100)
			}
		}
		_ = t
	}
	best = 1 << 62
	var total int64
	for run := 0; run < runs; run++ {
		start := time.Now()
		var t int64 = 0
		for y := int64(0); y < size; y++ {
			for x := int64(0); x < size; x++ {
				cr := (float64(x) - 100.0) / 50.0
				ci := (float64(y) - 100.0) / 50.0
				t += mandelbrotIter(cr, ci, 100)
			}
		}
		elapsed := time.Since(start).Nanoseconds()
		if elapsed < best {
			best = elapsed
			total = t
		}
	}
	fmt.Printf("Mandelbrot %dx%d:\n", size, size)
	fmt.Printf("  Time: %d ms\n", best/1000000)
	fmt.Printf("  Result: %d\n", total)

	// Benchmark 4: Distance computation
	n = 10000000
	// warmup
	{
		var s float64 = 0.0
		for i := int64(0); i < n; i++ {
			x := float64(i) * 0.001
			s += distance(x, 0.0, 0.0, x)
		}
		_ = s
	}
	best = 1 << 62
	for run := 0; run < runs; run++ {
		start := time.Now()
		var s float64 = 0.0
		for i := int64(0); i < n; i++ {
			x := float64(i) * 0.001
			s += distance(x, 0.0, 0.0, x)
		}
		elapsed := time.Since(start).Nanoseconds()
		if elapsed < best {
			best = elapsed
			resultF = s
		}
	}
	fmt.Printf("Distance (%d iterations):\n", n)
	fmt.Printf("  Time: %d ms\n", best/1000000)
	fmt.Printf("  Result: %.10f\n", resultF)

	// Benchmark 5: Trig computation
	n = 5000000
	trigSum(n) // warmup
	best = 1 << 62
	for run := 0; run < runs; run++ {
		start := time.Now()
		r := trigSum(n)
		elapsed := time.Since(start).Nanoseconds()
		if elapsed < best {
			best = elapsed
			resultF = r
		}
	}
	fmt.Printf("Trig sum(%d):\n", n)
	fmt.Printf("  Time: %d ms\n", best/1000000)
	fmt.Printf("  Result: %.10f\n", resultF)
}

File: benchmarks/float_compute.c
#include <stdio.h>
#include <stdint.h>
#include <math.h>
#include <time.h>

#define RUNS 3

// Basel problem: sum of 1/i^2 for i=1..n
double sum_inv_squares(int64_t n) {
    double sum = 0.0;
    for (int64_t i = 1; i <= n; i++) {
        double fi = (double)i;
        sum += 1.0 / (fi * fi);
    }
    return sum;
}

// Leibniz formula for pi: pi/4 = 1 - 1/3 + 1/5 - 1/7 + ...
double leibniz_pi(int64_t n) {
    double sum = 0.0;
    double sign = 1.0;
    for (int64_t i = 1; i <= n; i++) {
        double denom = (double)(i * 2 - 1);
        sum += sign / denom;
        sign = -sign;
    }
    return sum * 4.0;
}

// Mandelbrot escape iteration
int64_t mandelbrot_iter(double cr, double ci, int64_t max_iter) {
    double zr = 0.0, zi = 0.0;
    for (int64_t i = 0; i < max_iter; i++) {
        double tr = zr * zr - zi * zi + cr;
        zi = 2.0 * zr * zi + ci;
        zr = tr;
        if (zr * zr + zi * zi > 4.0) {
            return i;
        }
    }
    return max_iter;
}

// Euclidean distance using sqrt
double distance(double x1, double y1, double x2, double y2) {
    double dx = x2 - x1;
    double dy = y2 - y1;
    return sqrt(dx * dx + dy * dy);
}

// Sum of sin(x) + cos(x) for x = 0..n*0.001
double trig_sum(int64_t n) {
    double sum = 0.0;
    for (int64_t i = 0; i < n; i++) {
        double x = (double)i * 0.001;
        sum += sin(x) + cos(x);
    }
    return sum;
}

int64_t get_time_ns() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (int64_t)ts.tv_sec * 1000000000LL + (int64_t)ts.tv_nsec;
}

int main() {
    printf("=== C Float Benchmarks ===\n");

    // Benchmark 1: Basel problem
    int64_t n = 10000000;
    sum_inv_squares(n); // warmup
    int64_t best = INT64_MAX;
    double result_f;
    for (int run = 0; run < RUNS; run++) {
        int64_t start = get_time_ns();
        double r = sum_inv_squares(n);
        int64_t elapsed = get_time_ns() - start;
        if (elapsed < best) { best = elapsed; result_f = r; }
    }
    printf("Basel sum(1/i^2, 1..%ld):\n", n);
    printf("  Time: %ld ms\n", best / 1000000);
    printf("  Result: %.10f\n", result_f);

    // Benchmark 2: Leibniz pi
    n = 10000000;
    leibniz_pi(n); // warmup
    best = INT64_MAX;
    for (int run = 0; run < RUNS; run++) {
        int64_t start = get_time_ns();
        double r = leibniz_pi(n);
        int64_t elapsed = get_time_ns() - start;
        if (elapsed < best) { best = elapsed; result_f = r; }
    }
    printf("Leibniz pi(%ld terms):\n", n);
    printf("  Time: %ld ms\n", best / 1000000);
    printf("  Result: %.10f\n", result_f);

    // Benchmark 3: Mandelbrot grid
    int64_t size = 200;
    // warmup
    {
        int64_t t = 0;
        for (int64_t y = 0; y < size; y++)
            for (int64_t x = 0; x < size; x++)
                t += mandelbrot_iter(((double)x - 100.0) / 50.0, ((double)y - 100.0) / 50.0, 100);
    }
    best = INT64_MAX;
    int64_t total;
    for (int run = 0; run < RUNS; run++) {
        int64_t start = get_time_ns();
        int64_t t = 0;
        for (int64_t y = 0; y < size; y++) {
            for (int64_t x = 0; x < size; x++) {
                double cr = ((double)x - 100.0) / 50.0;
                double ci = ((double)y - 100.0) / 50.0;
                t += mandelbrot_iter(cr, ci, 100);
            }
        }
        int64_t elapsed = get_time_ns() - start;
        if (elapsed < best) { best = elapsed; total = t; }
    }
    printf("Mandelbrot %ldx%ld:\n", size, size);
    printf("  Time: %ld ms\n", best / 1000000);
    printf("  Result: %ld\n", total);

    // Benchmark 4: Distance computation
    n = 10000000;
    // warmup
    {
        double s = 0.0;
        for (int64_t i = 0; i < n; i++) {
            double x = (double)i * 0.001;
            s += distance(x, 0.0, 0.0, x);
        }
    }
    best = INT64_MAX;
    for (int run = 0; run < RUNS; run++) {
        int64_t start = get_time_ns();
        double s = 0.0;
        for (int64_t i = 0; i < n; i++) {
            double x = (double)i * 0.001;
            s += distance(x, 0.0, 0.0, x);
        }
        int64_t elapsed = get_time_ns() - start;
        if (elapsed < best) { best = elapsed; result_f = s; }
    }
    printf("Distance (%ld iterations):\n", n);
    printf("  Time: %ld ms\n", best / 1000000);
    printf("  Result: %.10f\n", result_f);

    // Benchmark 5: Trig computation
    n = 5000000;
    trig_sum(n); // warmup
    best = INT64_MAX;
    for (int run = 0; run < RUNS; run++) {
        int64_t start = get_time_ns();
        double r = trig_sum(n);
        int64_t elapsed = get_time_ns() - start;
        if (elapsed < best) { best = elapsed; result_f = r; }
    }
    printf("Trig sum(%ld):\n", n);
    printf("  Time: %ld ms\n", best / 1000000);
    printf("  Result: %.10f\n", result_f);

    return 0;
}
